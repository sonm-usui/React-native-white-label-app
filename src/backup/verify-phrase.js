// @flow

import React, { Component } from 'react'
import { Keyboard } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'

import type {
  VerifyRecoveryPhraseProps,
  VerifyRecoveryPhraseState,
  ReactNavigationBackup,
} from './type-backup'
import type { Store } from '../store/type-store'

import { HAS_VERIFIED_RECOVERY_PHRASE } from './type-backup'
import { CustomView, Icon, CustomHeader } from '../components'
import {
  verifyRecoveryPhraseRoute,
  exportBackupFileRoute,
  selectRecoveryMethodRoute,
} from '../common'
import { colors } from '../common/styles/constant'
import styles from './styles'
import {
  VERIFY_BACK_TEST_ID,
  VERIFY_CONTAINER_TEST_ID,
  VERIFY_INPUT_PLACEHOLDER,
} from './backup-constants'
import { pinHash as generateKey } from '../lock/pin-hash'
import VerifyPhrase from '../components/backup-restore-passphrase/backup-restore-passphrase'
import { getBackupPassphrase } from '../store/store-selector'
import { withStatusBar } from '../components/status-bar/status-bar'
import {
  hasVerifiedRecoveryPhrase,
  generateBackupFile,
} from '../backup/backup-store'
import { safeSet, walletSet } from '../services/storage'

const backImage = require('../images/icon_backArrow_white.png')

export class VerifyRecoveryPhrase extends Component<
  VerifyRecoveryPhraseProps,
  VerifyRecoveryPhraseState
> {
  state = {
    error: false,
  }

  static navigationOptions = ({
    navigation: { goBack },
  }: ReactNavigationBackup) => ({
    header: () => (
      <CustomHeader backgroundColor={colors.main} largeHeader flatHeader>
        <CustomView style={[styles.headerSpacer]}>
          <Icon
            medium
            onPress={() => goBack(null)}
            testID={VERIFY_BACK_TEST_ID}
            iconStyle={[styles.headerBackIcon]}
            src={backImage}
          />
        </CustomView>
      </CustomHeader>
    ),
    gestureEnabled: false,
    headerShown: true,
  })

  verifyRecoveryPhrase = async (passphrase: string) => {
    const cleanedPassphrase = passphrase
      .replace(/\s\s+/g, ' ')
      .replace(/(\r\n|\n|\r)/gm, ' ')
      .toLowerCase()
      .trim()
    const { recoveryPassphrase } = this.props
    const { initialRoute } = this.props.route.params

    const hashedPassphrase: string | null = await generateKey(
      cleanedPassphrase,
      this.props.recoveryPassphrase.salt
    )

    if (recoveryPassphrase.hash === hashedPassphrase) {
      if (this.props.isCloudBackupEnabled) {
        this.props.navigation.navigate(selectRecoveryMethodRoute, {
          initialRoute,
        })
      } else {
        this.props.hasVerifiedRecoveryPhrase()
        this.props.generateBackupFile()
        try {
          walletSet(HAS_VERIFIED_RECOVERY_PHRASE, 'true')
          safeSet(HAS_VERIFIED_RECOVERY_PHRASE, 'true')
        } catch (e) {
        } finally {
          this.props.navigation.navigate(exportBackupFileRoute, {
            initialRoute,
          })
        }
      }
      this.setState({ error: false })
    } else {
      this.setState({ error: true })
    }

    Keyboard.dismiss()
  }

  componentWillUnmount = () => {
    this.setState({ error: false })
  }

  render() {
    // TODO: Add error UI when that is designed
    return (
      <VerifyPhrase
        testID={VERIFY_CONTAINER_TEST_ID}
        placeholder={VERIFY_INPUT_PLACEHOLDER}
        onSubmit={this.verifyRecoveryPhrase}
        errorState={this.state.error}
        navigation={this.props.navigation}
      />
    )
  }
}

const mapStateToProps = (state: Store) => {
  return {
    recoveryPassphrase: getBackupPassphrase(state),
    isCloudBackupEnabled: false,
  }
}

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      hasVerifiedRecoveryPhrase,
      generateBackupFile,
    },
    dispatch
  )

export const verifyRecoveryPhraseScreen = {
  routeName: verifyRecoveryPhraseRoute,
  screen: withStatusBar({ color: colors.main })(
    connect(mapStateToProps, mapDispatchToProps)(VerifyRecoveryPhrase)
  ),
}
