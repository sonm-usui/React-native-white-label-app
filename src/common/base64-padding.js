// @flow

export function addBase64Padding(data: string): string {
  const trail = data.length % 4
  if (trail === 0) {
    return data
  }

  return `${data}${'='.repeat(4 - trail)}`
}

export function removeBase64Padding(data: string): string {
  return data.split('=')[0]
}
