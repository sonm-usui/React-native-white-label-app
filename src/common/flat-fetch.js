// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this function is just a wrapper on top of fetch.
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const acceptHeaders = { Accept: 'application/json;flavor=didcomm-msg' }

export async function flatFetch(
  url: string,
  body?: string,
  headers?: { [string]: any }
): Promise<[null | typeof Error, null | string, null | Response]> {
  const options = {
    mode: 'cors',
    ...(body ? { method: 'POST', body } : {}),
    ...(headers ? headers : {}),
  }
  try {
    const response = await fetch(url, options)
    const responseText = await response.text()
    // request redirected
    if (response && response.status === 302) {
      return [null, null, response]
    }
    if (!response.ok) {
      throw new Error(responseText)
    }
    return [null, responseText, response]
  } catch (e) {
    return [e, null, null]
  }
}
