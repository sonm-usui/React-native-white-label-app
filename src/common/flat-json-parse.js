// @flow

export function flatJsonParse(
  text: string
): [null | typeof Error, null | Object] {
  try {
    return [null, JSON.parse(text)]
  } catch (e) {
    return [e, null]
  }
}
