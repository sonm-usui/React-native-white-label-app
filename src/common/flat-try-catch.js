// @flow

export const flatTryCatch = (fn: (...variableArgs: any[]) => any) => (
  ...args: any[]
): [null | typeof Error, null | any] => {
  try {
    return [null, fn(...args)]
  } catch (e) {
    return [e, null]
  }
}
