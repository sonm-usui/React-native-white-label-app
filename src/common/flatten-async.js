// @flow

// If you look into the return type of this function
// you should notice that we are returning an array from  this function
// this is just a helper function to wrap on top of other async functions
// However, this changes a lot of ways we write code and handle errors
// By returning an array in which first item is an error
// and second item is the response
// We get two important things
// 1. We don't have to write try...catch statements everywhere
// 2. We are forced to handle errors. No more forgetting to add try...catch
// TODO:KS Write more details about both 1 & 2
// Here are the problems that this function aims to solve
// Problem #1 leads to try...catch nesting hell if there are multiple async
// calls
// Problem #2 leads to unexpected app crashes at runtime

export const flattenAsync = (
  fn: (...variableArgs: any[]) => Promise<any>
) => async (
  ...args: Array<any>
): Promise<[null | typeof Error, null | any]> => {
  try {
    const response = await fn(...args)
    return [null, response]
  } catch (e) {
    return [e, null]
  }
}
