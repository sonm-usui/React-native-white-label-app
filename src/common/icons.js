// @flow
import React from 'react'
import { Icon } from 'react-native-eva-icons'
import { colors } from './styles/constant'
import { moderateScale } from 'react-native-size-matters'

type IconProps = {
  width?: number,
  height?: number,
  color?: string,
}

type CommonIconProps = {
  name: string,
} & IconProps

export const HOME_ICON = 'home-outline'
export const CONNECTIONS_ICON = 'people-outline'
export const SETTINGS_ICON = 'settings-2-outline'

export const ERROR_ICON = 'alert-circle-outline'
export const CAMERA_ICON = 'camera-outline'
export const CHECK_MARK_ICON = 'checkmark-circle-2-outline'
export const HOME_MENU_ICON = 'menu-outline'
export const ANDROID_BACK_ARROW_ICON = 'arrow-back-outline'
export const IOS_BACK_ARROW_ICON = 'arrow-ios-back-outline'
export const MORE_ICON = 'more-vertical-outline'
export const DELETE_ICON = 'trash-2-outline'
export const CLOSE_ICON = 'close-outline'
export const CHAT_ICON = 'message-square-outline'
export const INFO_ICON = 'info-outline'
export const ARROW_RIGHT_ICON = 'chevron-right-outline'
export const LOCK_ICON = 'lock'
export const SAVE_ICON = 'save-outline'
export const ALERT_ICON = 'alert-circle-outline'
export const ARROW_FORWARD_ICON = 'arrow-ios-forward-outline'
export const CHECKMARK_ICON = 'checkmark-outline'
export const SHARE_ICON = 'share-outline'
export const GRID_ICON = 'grid-outline'

// common icon class is implemented to set default values (except name) for icons
// in that case we should only set icon name and get an icon with default color and size
export const EvaIcon = (props: CommonIconProps) => {
  const { width, height, color } = props

  return (
    <Icon
      {...props}
      width={width ? width : moderateScale(24)}
      height={height ? height : moderateScale(24)}
      fill={color ? color : colors.gray2}
    />
  )
}
