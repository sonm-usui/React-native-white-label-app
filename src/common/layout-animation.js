// @flow

import { LayoutAnimation } from 'react-native'
import { checkIfAnimationToUse } from '../bridge/react-native-cxs/RNCxs'

export function animateLayout() {
  if (!checkIfAnimationToUse()) {
    LayoutAnimation.configureNext(
      LayoutAnimation.create(
        150,
        LayoutAnimation.Types.linear,
        LayoutAnimation.Properties.opacity
      )
    )
  }
}
