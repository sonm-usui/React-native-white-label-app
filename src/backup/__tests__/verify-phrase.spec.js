// @flow

import React from 'react'
import renderer from 'react-test-renderer'
import { Provider } from 'react-redux'
import { VerifyRecoveryPhrase } from '../verify-phrase'
import { getStore } from '../../../__mocks__/static-data'
import { settingsRoute } from '../../common'

jest.mock('react-native-keyboard-aware-scroll-view')


describe('<VerifyRecoveryPhrase />', () => {
  const recoveryPassphrase = 'hello some passphrase'

  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.runOnlyPendingTimers()
  })

  const navigation = {
    navigate: jest.fn(),
    goBack: jest.fn(),
    setParams: jest.fn(),
  }
  const route = {
    params: {
      recoveryPassphrase,
      initialRoute: settingsRoute,
      navigateBack: jest.fn(),
    },
  }

  const mockStoreData = getStore()

  it('should match snapshot', () => {
    const tree = renderer
      .create(
        <Provider store={mockStoreData}>
          <VerifyRecoveryPhrase
            recoveryPassphrase={{
              phrase: recoveryPassphrase,
              salt: 'salt',
              hash: 'hash',
            }}
            navigation={navigation}
            route={route}
            hydrateCloudBackup={jest.fn()}
            submitPassphrase={() => {}}
            restoreStatus={() => {}}
            resetError={() => {}}
            error={false}
            status="none"
            isCloudBackupEnabled={false}
            generateBackupFile={() => {}}
            hasVerifiedRecoveryPhrase={() => {}}
          />
        </Provider>
      )
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
