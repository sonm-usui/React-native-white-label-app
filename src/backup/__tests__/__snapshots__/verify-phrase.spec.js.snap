// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<VerifyRecoveryPhrase /> should match snapshot 1`] = `
<RCTSafeAreaView
  accessibilityLabel="verify-passphrase-container-container-touchable"
  accessible={false}
  emulateUnlessSupported={true}
  focusable={true}
  hitSlop={
    Object {
      "bottom": 20,
      "left": 20,
      "right": 20,
      "top": 20,
    }
  }
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  safeArea={true}
  style={
    Object {
      "backgroundColor": "#86B93B",
      "flex": 1,
    }
  }
  testID="verify-passphrase-container-container-touchable"
>
  <Image
    source={
      Object {
        "default": "RelativeImageStub",
      }
    }
    style={
      Array [
        Object {
          "flex": 1,
          "position": "absolute",
        },
      ]
    }
  />
</RCTSafeAreaView>
`;
